core/shared/src/main/scala/zio/internal/FiberContext.scala

```diff
@@ -654,31 +653,29 @@ private[zio] final class FiberContext[E, A](
     }
 
   @tailrec
-  private[this] final def enterAsync(): Boolean = {
+  private[this] final def enterAsync(zio: ZIO[_, _, _]): Boolean = {
     val oldState = state.get
 
     oldState match {
       case Executing(_, observers, interrupt) =>
-        val newState = Executing(FiberStatus.Suspended(interruptible), observers, interrupt)
+        val interruptible = this.interruptible
+        val newState      = Executing(FiberStatus.Suspended(interruptible, zio), observers, interrupt)
 
-        if (!state.compareAndSet(oldState, newState)) enterAsync()
-        else if (shouldInterrupt) {
-          // Fiber interrupted, so go back into running state:
-          exitAsync()
-          false
-        } else true
+        if (interruptible && interrupt) false
+        else if (!state.compareAndSet(oldState, newState)) enterAsync(zio)
+        else true
 
       case _ => false
     }
   }
 
   @tailrec
-  private[this] final def exitAsync(): Boolean = {
+  private[this] final def exitAsync(zio: ZIO[_, _, _]): Boolean = {
     val oldState = state.get
 
     oldState match {
-      case Executing(FiberStatus.Suspended(_), observers, interrupt) =>
-        if (!state.compareAndSet(oldState, Executing(FiberStatus.Running, observers, interrupt))) exitAsync()
+      case Executing(FiberStatus.Suspended(_, oldZio), observers, interrupt) if zio eq oldZio =>
```

Comment:
This assumes all ZIO values have unique references which is not true in general (consider a ZIO value stored in a `val` and reused), I'm not sure if that's relevant, but if it is, would it make sense to use a volatile counter?

Reply:
You are right.
Now I think a volatile counter is a better solution.
Another solution: pass `new Object` instead of `curZio` as sync pointer. But it will slightly increase allocation rate.